[REVIEW] NEXUS-5511: Improve not found reasoning

cstamas opened the pull request:
Nexus might response with 404 Not Found in for many reasons, and proxy404 will be "just another" one of them.

Nexus since 2.2 (or 2.3?) bubbles up the actual not found message to REST layer and HTML content page even renders the message to user. But, unfortunately, the message was a constant ("The item not found in ...").

These changes now changes critical places like NFC, local storage, remote storage, processors to give more hints why response is 404.

Aside improving messages, this pull also gives ability to "hint" reason upfront, like the case with localOnly or groupLocalOnly flags is (by setting these flags you still don't know the outcome of the response). A similar technique would be used in proxy404 to tell user that "Whitelist rejected to perform a remote request".

cstamas committed:
INF cleanup: WIP

cstamas committed:
Merge remote-tracking branch 'origin/master' into item-not-found-cleanup

cstamas committed:
NEXUS-5511: Improve reasoning for INFEx

Implemented main cases.

cstamas committed:
NEXUS-5511: Improving some messages.

cstamas committed:
NEXUS-5511: Fixed broken messages.

Messages in two INFEx were constructed wrong.
Also, UT expectations fixed, aligned with actual message.

cstamas committed:
NEXUS-5511: Slipped UT for expectation fix

cstamas committed:
NEXUS-5511: Undoing indexer manager, as it has formatting issues.

cstamas committed:
NEXUS-5511: Cleanup

cstamas committed:
NEXUS-5511: Cleanup

cstamas commented:
CI:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF22

peterlynch commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundReasons.java:
```diff
@@ -0,0 +1,179 @@
+ * @author cstamas
+ * @since 2.4
+ */
+public class ItemNotFoundReasons
```
abstract?

cstamas replied:
Added private constructor to prevent instantiation.

peterlynch commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/router/DefaultRepositoryRouter.java:
```diff
@@ -519,13 +528,14 @@ public RequestRoute getRequestRouteForRequest( ResourceStoreRequest request )
                 {
                     // this is not the main facet or the repo is not exposed
-                    throw new ItemNotFoundException( request );
+                    throw new ItemNotFoundException( reasonFor( request, "Repository %s exists but is not exposed!",
```
I could live without the exclamation marks ;)

cstamas replied:
Fixed

peterlynch commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSPeer.java:
```diff
@@ -482,7 +496,7 @@ protected void handleRenameOperation( File hiddenTarget, File target )
             {
                 getLogger().info( "Rename operation succeeded after {} retries on {} --> {}",
-                                  new Object[]{ i, hiddenTarget.getAbsolutePath(), target.getAbsolutePath() } );
+                    new Object[] { i, hiddenTarget.getAbsolutePath(), target.getAbsolutePath() } );
```
I think var args make new Object[] not required if memory serves. Same in several other places.

cstamas replied:
These lines were not touched by me, it was only reformat doing it. Updating the SLF4J use to the level where new method was introduced on it would require more effort that this here anyway.

jdillon replied:
This is generally pretty easy to clean up... and we probably should try to clean this stuff up as we go along if we don't plan to fix all logging usage in one go (which I'm not recommending).

peterlynch commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -25,41 +25,89 @@
-        super( "Item not found on path " + path, cause );
-        this.repository = null;
-        this.request = null;
+        super( reason.getMessage(), cause ); // will NPE with null reason
```
should add @throws NullPointerException to related javadoc imo

cstamas replied:
Reworking this a bit, initial solution was awkward.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
NEXUS-5511: Applied comments and simplified

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
NEXUS-5511: Not quite related.

Using SLF4J's new vararg method for logging

cstamas commented:
Updated logging to new SLF4J api in DefaultFSPeer.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -25,41 +29,129 @@
+     * {@link ItemNotFoundInRepositoryReason} instead. This one is used in places like {@link RepositoryRouter}, where
+     * the "targeted" repository is still unknown or similar places.
+     */
+    public static class ItemNotFoundReason
```
This is subjective, but I find it odd that standalone ItemNotFoundReasons creates instances of nested ItemNotFoundException.ItemNotFoundReason. I would probably collapse ItemNotFoundReasons onto ItemNotFoundException, but there are other sensible alternatives too.

cstamas replied:
INFReasons class removed.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -1320,6 +1329,9 @@ protected boolean checkPostConditions( final ResourceStoreRequest request, final
                 {
+                    request.addItemNotFoundReason( reasonFor( request, this,
+                        "Retrieval of request %s in repository %s prevented.", request.getRequestPath(),
+                        RepositoryStringUtils.getHumanizedNameString( this ) ) );
```
What additional information does the message above provide? Either add info about processor, if it is relevant to the user, or don't provide additional info.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractProxyRepository.java:
```diff
@@ -1184,6 +1195,9 @@ protected StorageItem doRetrieveItem0( ResourceStoreRequest request, AbstractSto
                 {
                     // escape
+                    request.addItemNotFoundReason( ItemNotFoundReasons.reasonFor( request, this,
+                        "Proxying of path %s in repository %s prevented", request.getRequestPath(),
```
Same as above, this does not provide any additional info. Either make it useful or remove it.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractProxyRepository.java:
```diff
@@ -263,11 +267,8 @@ public boolean expireProxyCaches( final ResourceStoreRequest request, final Walk
-                )
-            );
+                new RepositoryEventExpireProxyCaches( this, request.getRequestPath(),
+                    request.getRequestContext().flatten(), expireCacheWalkerProcessor.isCacheAltered() ) );
```
In the future, please push formatting changes as separate commits. When part of a larger unrelated pull, formatting changes make reviews harder than they should be.

ifedorenko commented:
-0

I agree with the general approach, but I am concerned about using ResourceStoreRequest to capture ItemNotFoundReason. Ideally, requests should be immutable objects that describe original client requests to the Repository/Storage subsystem. We already use requests to track request execution state, i.e. push/pop path, applied mappings, etc, which I believe is wrong. With this proposed change ResourceStoreRequest will also track request execution results, which will make the problem even worse. 

As far as I can tell AbstractRepository.checkPostConditions and AbstractRepository.checkRequestProcessors are the only two places where the code really needs to collect info about conditions not met, everywhere else we can just throw ItemNotFoundException directly. I suggest we find a cleaner way to communicate info about failing conditions.

nabcos committed:
Merge remote-tracking branch 'upstream/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
NEXUS-5511: Simplified

Changes:
* Removed not found reasons class, molded with INFex
* Reason now holds a "copy and detached" copy of request (static but mutable copy)
* ResourceStoreRequest does NOT track INF reasons anymore
* applied other comments, like removed INF reasoning with processors (processor will throw if needs that)

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF22-4

cstamas commented:
@ifedorenko I'd like you to give this pull a 2nd pass (as your comments were applied while you were on vacation)

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -1303,6 +1311,8 @@ protected boolean checkRequestProcessors( final ResourceStoreRequest request, fi
                 {
+                    // let's not state anything about reason, and allow processor say why it rejected
+                    // by throwing appropriate exception
                     return false;
```
This is not good enough. RequestProcessor needs a way to provide specific reasons why the request was rejected all the way back to the user. And no, request object is not the right place to store this additional information ;-)

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -1320,6 +1330,8 @@ protected boolean checkPostConditions( final ResourceStoreRequest request, final
                 if ( !processor.shouldRetrieve( this, request, item ) )
                 {
+                    // let's not state anything about reason, and allow processor say why it rejected
+                    // by throwing appropriate exception
```
Same as above, RequestProcessor needs a way to communicate reasons back to the user.

ifedorenko commented:
-1. RequestProcessor's need a way to communicate reasons request were rejected back to the user.

rseddon commented:
If we can't cover the  cases mentioned by Igor before the 2.4 release is done I'd suggest merging this pull request anyhow, and filing a separate bug for those.  What we have currently makes it practically impossible for users to self-diagnose 404's, and this causes a lot of support tickets.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

Conflicts:
	nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/AbstractMavenRepository.java
	nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/ChecksumContentValidator.java
	nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java

cstamas committed:
NEXUS-5511: Improve INFEx reasoning.

Introducing new RequestProcessor2 iface to completely replace
existing RequestProcessor's. New iface has means to provide
"proper reasoning" why something should not be found.

Also, unlike old iface, new one is able to interfere with INFex
only, not with other (usually problem meaning) exceptions
like IllegalOperationEx or AccessDeniedEx.

An adapter is provided to wrap old implementations if needed,
but cleanest way for RequestProcessor implementors would be to
reimplement their code agains new iface.

cstamas commented:
CI job:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF38-1

jdillon commented on nexus-core/pom.xml:
```diff
@@ -58,6 +58,10 @@
     <dependency>
       <groupId>org.sonatype.sisu.goodies</groupId>
+      <artifactId>goodies-common</artifactId>
+    </dependency>
```
This may have a bit more impact than this change.  Pluginkit and staging, maybe some others (have to check) also include goodies-common, so we'll have to update those too.

But FTR I'm pro adding goodies-common to core.

cstamas replied:
Given that introducing goodies to core is not quite simple (as it has implications for aforementioned plugins at least), maybe it deserves a separate (cleaner and simpler to review) pull first?

jdillon replied:
yes probably

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -24,42 +29,206 @@
+     * @param request
+     * @param repository
+     * @param message
+     * @param params
```
Why do you feel you need to put empty javadoc tags which do not provide any other information?  This is pointless and only increases the chances that when a parameter is renamed that the docs will be out of sync (which I've seen in many many many places inside of NX).

General rule to follow, if you aren't going to actually add any documentation... then do not add empty javadoc tags.  Its not like w/o these tags that javadoc won't actually report this method has parameters.  This is just extra bytes/lines of code which add zero value.  Please avoid.

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -24,42 +29,206 @@
+         * 
+         * @return the reason message.
+         */
+        public String getMessage()
```
This is a good example of bad documentation.  The docs here add zero value, as any sane developer should assume a "String getMessage()" method would return a message.  Documentation would really only serve here to explain any reason why it wouldn't return a string.  And either way, there is no reason you document @return here as the method description covers it.  This is just extra wasted bytes to maintain which add zero value.

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -94,6 +265,7 @@ public ItemNotFoundException( final ResourceStoreRequest request, final Reposito
      * @param cause
+     * @deprecated Use constructor with {@link ItemNotFoundReason} instead.
      */
     public ItemNotFoundException( final ResourceStoreRequest request, final Repository repository, final Throwable cause )
```
@Deprecate and @deprecate should come in pairs... the later to explain why the former was used.

jdillon commented:
-0.333 for lack of @since on new public api

-0.333 for the poor use of javadoc

-0.334 for the formatting changes mixed in which made this even more difficult to comprehend the actual changes

:-P

cstamas committed:
NEXUS-5511: Applied comments.

Javadoc fixes and removed unused method.

cstamas committed:
NEXUS-5511: removed unused imports.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

Conflicts:
	nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSLocalRepositoryStorage.java
	nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSPeer.java

cstamas committed:
NEXUS-5511: Simplification and method renames

Simplified RequestProcessor2 by removing a not used
call shouldCache. This would not even work properly
as would circumvent retries and other logic from core

Methods on RequestProcessor2 renamed to better reflect
when they are called.

cstamas committed:
NEXUS-5511: Revamped request processor API

Backed off from initial design, and returned to throwing an
exception when some signaling is needed from processor.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
NEXUS-5511: Renamed newly introduced processor

Tried to find better name, RequestStrategy instead of
RequestProcessor2.

Also, real backward compatibility wrt old RequestProcessor
restored, as previous changes were breaking it. Now a
"deprecated" support should result in old code running
unchanged.

cstamas committed:
NXCM-5511: Bugfix

Checking are strategies needed at all to be
invoked was broken, old code mingled with new.

cstamas committed:
NEXUS-5511: Updated since tags

Bumped to 2.5

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractProxyRepository.java:
```diff
@@ -976,19 +975,6 @@ protected DownloadMirrorSelector openDownloadMirrorSelector( ResourceStoreReques
-
-            if ( !shouldCache )
-            {
-                return item;
```
Did you remove this on purpose or by accident? If on purpose, why?

cstamas replied:
Yes, as newly introduced RequestStrategy does not have this method anymore, and I did not want to leave RequestProcessors (those deprecated ones) still littered here.

Also, as I checked, no implementation of RequestProcessor was doing anything else with this method, they never returned `false` here.

Finally, even if they _would return false_ here, I am completely sure it would just cause problems, similar to https://github.com/sonatype/nexus/pull/820 HTTP connection leak, as the RRS items (those wrapping HTTP response body) would bubble up all way to the REST layer, where in case of a `?describe` requests -- and sure for any non-200 OK response too -- ContentLocators are just tossed away ( = leak same as NEXUS-5468 in fixed in pull above).

ifedorenko commented:
Still not there but definitely getting closer. 

Before I can agree with this change,
- I want to understand if behaviour change in AbstractProxyRepository was on purpose or not (see inline question above)
- Decide if we want to maintain backwards compatibility for RequestProcessor, I personally think it's not worth it, but lets see what others say about it
- Undo all formatting-only changes to make this pull smaller and easier to review. The easiest way to do this is probably squash all commits into one, then undo all formatting-only changes and amend, but feel-free to do compensating commit.

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5511-improve-infex-reasoning

Conflicts:
	nexus-core/src/main/java/org/sonatype/nexus/proxy/storage/local/fs/DefaultFSPeer.java

ifedorenko commented:
-1

As discussed in email, there is no need to provide backwards compatibility for existing RequestProcessor implementations. Please replace  RequestProcessor with RequestStrategy and remove code related to RequestProcessor backwards compatibility.

cstamas committed:
NEXUS-5511: Removing RequestProcess completely

cstamas committed:
NEXUS-5511: Optimizing imports

cstamas committed:
NEXUS-5511: Fixing signature of support class

As it was not bringing in throws from iface...

cstamas committed:
Fixing compilation in UT after sign fix

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/ItemNotFoundException.java:
```diff
@@ -25,42 +30,200 @@
+
     /**
-     * Do not use this constructor!
+     * Constructor.
```
I think the comment should read something like

   Constructor of class ItemNotFoundException written in java programming language

(seriously, though, please either provide meaningful javadoc or don't provide any at all)

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -267,14 +275,44 @@ public RepositoryTaskFilter getRepositoryTaskFilter()
+        synchronized ( requestStrategies )
         {
-            requestProcessors = new HashMap<String, RequestProcessor>();
+            return Maps.newHashMap( requestStrategies );
```
Need to return unmodifiable map

cstamas replied:
It returns a _copied_ (modifiable) map, a detached "snapshot" of strategies in a moment of method invocation. Javadoc clearly states that:
https://github.com/sonatype/nexus/pull/726/files#L14R381

Unsure why you think it need unmodifiable map...

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -267,14 +275,44 @@ public RepositoryTaskFilter getRepositoryTaskFilter()
+     * 
+     * @return
+     */
+    protected Map<String, RequestStrategy> getEffectiveRequestStrategies()
```
This method is now redundant, please remove it.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/AbstractRepository.java:
```diff
@@ -1287,40 +1310,33 @@ protected boolean checkConditions( ResourceStoreRequest request, Action action )
     {
-        if ( getRequestProcessors().size() > 0 )
+        final Map<String, RequestStrategy> effectiveRequestStrategies = getEffectiveRequestStrategies();
+        if ( !effectiveRequestStrategies.isEmpty() )
```
Why do you need !isEmpty check here? It only adds noise, for-each loop handles empty collections already.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/RequestStrategy.java:
```diff
@@ -0,0 +1,80 @@
+     * @throws IllegalOperationException
+     */
+    void onHandle( Repository repository, ResourceStoreRequest request, Action action )
+        throws ItemNotFoundException, IllegalOperationException;
```
It is not clear in what cases this method will be called and it expected behaviour if "item should not be handled", i.e. the difference between ItemNotFoundException and IllegalOperationException needs to be explained.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/RequestStrategy.java:
```diff
@@ -0,0 +1,80 @@
+     * @throws IllegalOperationException
+     */
+    void onServing( Repository repository, ResourceStoreRequest request, StorageItem item )
+        throws ItemNotFoundException, IllegalOperationException;
```
It is not clear in what cases IllegalOperationException should be thrown. Are there reasons to have this exception here at all? This method either allows serving of the artifact or throws ItemNotFoundException that explains the reason why the artifact should not be served.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/RequestStrategy.java:
```diff
@@ -0,0 +1,80 @@
+     * @throws IllegalOperationException
+     */
+    void onRemoteAccess( ProxyRepository repository, ResourceStoreRequest request )
+        throws ItemNotFoundException, IllegalOperationException;
```
Same as above, IllegalOperationException appears redundant.

ifedorenko commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/repository/RequestStrategy.java:
```diff
@@ -0,0 +1,80 @@
+ * @author cstamas
+ * @since 2.5
+ */
+public interface RequestStrategy
```
Why did you decide to rename RequestProcessor to RequestStrategy? Both names are too generic to explain the purpose of this interface, so I am not sure why change one poor name with another equally poor name.

cstamas replied:
Well, I did not rename RequestProcessor (at least not in events sequence) on purpose, it is the result of recent email exchanges we had (implicitly "gained"). 

The decision to _remove/break_ RequestProcessor happened today morning (my time), but before that decision, I had introduced RequestProcessor2 (initially very badly named, on Wed I renamed it to RequestStrategy) in parallel with keeping but deprecating ReqProcs.

As there was multiple problems with RP interface (badly named methods, inconsistent method signatures and such), but also how they got registered and unregistered to repository (was not thread safe and too error prone), I decided to introduce a "clean slate" _new_ RequestProcessor2 (again, later renamed to Strategy), while keeping backward compat (again, all this happened earlier this week, way before the email for 1/2/3 options was fired on list).

Now, that I applied the option 3, and removed RequestProcessors and support for backward compat, I can rename the RequestStrategy back to RequestProcessor if we see that change as really needed... I did not see it important. Either way, we broke backward compat.

jdillon replied:
names are not horrible... not sure there is anything better, at least I can't think of anything better atm.

ifedorenko commented:
-1

See few smaller/cosmetic comments inline

cstamas committed:
NEXUS-5511: Applied Igor's comments.

Changes:
* javadoc and since fixes, clearing up things
* removed redundant code and methods

ifedorenko commented:
+1
